File: apps/desktop/src/catalog-deletion.ts
import type { DeleteGlobalCatalogEntryRequest } from "./api";
import type { ManagedSettingsSnapshot } from "./types";

export type DeletableCatalogKind = "model" | "provider";

/** Display dependencies from the saved snapshot; native code rechecks on deletion. */
export function catalogDeletionBlockers(
  snapshot: ManagedSettingsSnapshot,
  kind: DeletableCatalogKind,
  resourceId: string,
): string[] {
  const global = snapshot.globalConfiguration;
  const profiles = new Set(
    kind === "provider"
      ? global.providers
          .find((entry) => entry.id === resourceId)
          ?.revisions.map(({ value }) => value.profile)
      : [],
  );
  const blockers = new Set<string>();
  for (const model of global.models.filter((entry) => !entry.archived)) {
    const current = model.revisions.find(
      ({ revision }) => revision === model.currentRevision,
    );
    if (current && profiles.has(current.value.providerProfile)) {
      blockers.add(`Model ${model.label}`);
    }
  }
  for (const space of snapshot.spaces) {
    const referenced = Object.entries(
      space.configuration.catalogRevisions,
    ).some(([key, reference]) => {
      if (key.startsWith(`${kind}:`) && reference.resourceId === resourceId)
        return true;
      const model = key.startsWith("model:")
        ? global.models
            .find((entry) => entry.id === reference.resourceId)
            ?.revisions.find(({ revision }) => revision === reference.revision)
        : undefined;
      return model !== undefined && profiles.has(model.value.providerProfile);
    });
    if (referenced)
      blockers.add(
        `Workspace ${space.name}${space.archived ? " (archived)" : ""}`,
      );
  }
  return [...blockers].sort();
}

/** Browser fixture counterpart to the native settings mutation. */
export function deleteCatalogEntryFixture(
  snapshot: ManagedSettingsSnapshot,
  kind: DeletableCatalogKind,
  request: DeleteGlobalCatalogEntryRequest,
): ManagedSettingsSnapshot {
  if (snapshot.globalConfiguration.revision !== request.expectedRevision) {
    throw new Error(
      "Global settings changed in another window. Reload and review the latest revision.",
    );
  }
  const key = kind === "model" ? "models" : "providers";
  if (
    !snapshot.globalConfiguration[key].some(
      (entry) => entry.id === request.resourceId,
    )
  ) {
    throw new Error(`The ${kind} is unknown.`);
  }
  if (catalogDeletionBlockers(snapshot, kind, request.resourceId).length) {
    throw new Error(
      `This ${kind} is still in use. Update its references first.`,
    );
  }
  const next = structuredClone(snapshot);
  const global = next.globalConfiguration;
  const defaults = global.defaults.revisions.find(
    ({ revision }) => revision === global.defaults.currentRevision,
  )!;
  global.revision += 1;
  global.defaults.currentRevision = global.revision;
  global.defaults.revisions.push({
    revision: global.revision,
    value: structuredClone(defaults.value),
  });
  if (kind === "model")
    global.models = global.models.filter(
      (entry) => entry.id !== request.resourceId,
    );
  else
    global.providers = global.providers.filter(
      (entry) => entry.id !== request.resourceId,
    );
  for (const space of next.spaces) {
    if (
      space.configuration.acceptedGlobalRevision === request.expectedRevision
    ) {
      space.configuration.acceptedGlobalRevision = global.revision;
    } else {
      space.pendingGlobalRevision = global.revision;
      space.status = "update_available";
      space.statusMessage = `Global revision ${global.revision} is ready to review and apply.`;
    }
  }
  return next;
}

File: apps/desktop/src-tauri/src/managed_configuration_commands/catalog_deletion.rs
//! Deletion of unused provider and model catalog entries, including their revisions.

use super::{
    AppState, CommandErrorDto, DesktopSettings, ManagedSettingsSnapshotDto, State,
    advance_unaffected_spaces, bump_global_revision, connect_guard, current_value,
    ensure_global_revision, settings_store, snapshot,
};
use serde::Deserialize;
use std::collections::BTreeSet;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct DeleteGlobalCatalogEntryInput {
    expected_revision: u64,
    resource_id: String,
}

#[derive(Clone, Copy, Debug)]
enum CatalogKind {
    Model,
    Provider,
}

impl CatalogKind {
    const fn name(self) -> &'static str {
        match self {
            Self::Model => "model",
            Self::Provider => "provider",
        }
    }

    const fn prefix(self) -> &'static str {
        match self {
            Self::Model => "model:",
            Self::Provider => "provider:",
        }
    }
}

#[tauri::command(rename_all = "camelCase")]
pub(crate) async fn delete_global_model(
    state: State<'_, AppState>,
    request: DeleteGlobalCatalogEntryInput,
) -> Result<ManagedSettingsSnapshotDto, CommandErrorDto> {
    delete_entry(state, request, CatalogKind::Model).await
}

#[tauri::command(rename_all = "camelCase")]
pub(crate) async fn delete_global_provider(
    state: State<'_, AppState>,
    request: DeleteGlobalCatalogEntryInput,
) -> Result<ManagedSettingsSnapshotDto, CommandErrorDto> {
    delete_entry(state, request, CatalogKind::Provider).await
}

async fn delete_entry(
    state: State<'_, AppState>,
    request: DeleteGlobalCatalogEntryInput,
    kind: CatalogKind,
) -> Result<ManagedSettingsSnapshotDto, CommandErrorDto> {
    let _guard = connect_guard(&state)?;
    let store = settings_store()?;
    let mut settings = store.load()?;
    apply_catalog_deletion(&mut settings, &request, kind)?;
    store.save(&settings)?;
    snapshot(state.inner(), &settings).await
}

fn apply_catalog_deletion(
    settings: &mut DesktopSettings,
    request: &DeleteGlobalCatalogEntryInput,
    kind: CatalogKind,
) -> Result<(), CommandErrorDto> {
    ensure_global_revision(settings, request.expected_revision)?;
    let global = &settings.global_configuration;
    let exists = match kind {
        CatalogKind::Model => global
            .models
            .iter()
            .any(|entry| entry.id == request.resource_id),
        CatalogKind::Provider => global
            .providers
            .iter()
            .any(|entry| entry.id == request.resource_id),
    };
    if !exists {
        return Err(CommandErrorDto::invalid(
            "resourceId",
            &format!("The {} is unknown.", kind.name()),
        ));
    }
    let mut dependents = BTreeSet::new();
    let provider_profiles = global
        .providers
        .iter()
        .filter(|entry| matches!(kind, CatalogKind::Provider) && entry.id == request.resource_id)
        .flat_map(|entry| &entry.revisions)
        .map(|revision| revision.value.profile.as_str())
        .collect::<BTreeSet<_>>();
    // Current model definitions must remain usable even before a workspace selects them.
    for entry in global.models.iter().filter(|entry| !entry.archived) {
        if current_value(entry)
            .is_some_and(|model| provider_profiles.contains(model.provider_profile.as_str()))
        {
            dependents.insert(format!("Model {}", entry.label));
        }
    }
    for space in &settings.spaces {
        let referenced = space
            .configuration
            .catalog_revisions
            .iter()
            .any(|(key, reference)| {
                if key.starts_with(kind.prefix()) && reference.resource_id == request.resource_id {
                    return true;
                }
                // A workspace may still use an older model revision after its provider changes.
                key.starts_with("model:")
                    && global
                        .models
                        .iter()
                        .find(|entry| entry.id == reference.resource_id)
                        .and_then(|entry| {
                            entry
                                .revisions
                                .iter()
                                .find(|revision| revision.revision == reference.revision)
                        })
                        .is_some_and(|revision| {
                            provider_profiles.contains(revision.value.provider_profile.as_str())
                        })
            });
        if referenced {
            dependents.insert(format!(
                "Workspace {}{}",
                space.display_name,
                if space.archived { " (archived)" } else { "" }
            ));
        }
    }
    if !dependents.is_empty() {
        return Err(CommandErrorDto::invalid(
            "resourceId",
            &format!(
                "This {} is still used by: {}. Update or remove those references and apply workspace changes before deleting it. Restore archived workspaces first.",
                kind.name(),
                dependents.into_iter().collect::<Vec<_>>().join(", ")
            ),
        ));
    }
    let previous_revision = global.revision;
    // Validate the revision increment before removing any saved configuration.
    bump_global_revision(&mut settings.global_configuration)?;
    match kind {
        CatalogKind::Model => settings
            .global_configuration
            .models
            .retain(|entry| entry.id != request.resource_id),
        CatalogKind::Provider => settings
            .global_configuration
            .providers
            .retain(|entry| entry.id != request.resource_id),
    }
    advance_unaffected_spaces(settings, previous_revision, &BTreeSet::new());
    Ok(())
}

#[cfg(test)]
mod tests;
